### PR TITLE
#184347020 Build out the user login endpoint

### PR DESCRIPTION
**Pull request by irakozetony**
### What does this PR do?

Creates the login with email & password endpoint

### Description of Task to be completed?

- [x] Implement login with email & password

### How should this be manually tested?

1. Clone [repo](https://github.com/atlp-rwanda/vikings-ec-bn.git)
2. Change into the directory vikings-ec-bn
3. Checkout to this branch: `feature/user-login-184347020`
4. In your `.env` file, create a `SECRET_TOKEN` variable with a value in it
5. Run the command `npm install` to install all packages
6. Run `npm run seeding` to rest migration and seeding
7. Go to the `http://localhost:<port>/api-docs/#/Authentication/post_api_v1_users_login` URL in your browser or `http://localhost:<port>/api/v1/users/login` in Postman

### What are the relevant pivotal tracker stories?

[#184347020](https://www.pivotaltracker.com/story/show/184347020)

### CODE REVIEW

**Review by Fidela1 — APPROVED** [review]

**Review by ndatumuremyi — APPROVED** [review]

**Review by Munezeroolivierhugue — APPROVED** [review]

**Review by fabmukunzi — APPROVED** [review]

File: src/controllers/user.controller.js [hunk 1]
```diff
@@ -1,22 +1,47 @@
...
+				error: err.message,
+				message: 'Failed to register a new user',
+			});
+		}
+	}
+
+	static async loginUser(req, res) {
+		try {
+			const user = await User.findOne({
+				where: { email: req.body.email },
```
Comment (ncutixavier) [inline src/controllers/user.controller.js]: Use destructuring assignment for email & password

File: src/controllers/user.controller.js [hunk 2]
```diff
@@ -1,22 +1,47 @@
...
+			});
+			const isValidPassword = await BcryptUtility.verifyPassword(
+				req.body.password,
+				user.password
+			);
+			if (!isValidPassword)
+				return res.status(400).json({ error: 'Wrong credentials' });
+
```
Comment (ncutixavier) [inline src/controllers/user.controller.js]: Put this in middleware

File: src/controllers/user.controller.js [hunk 3]
```diff
@@ -1,22 +1,47 @@
...
+			const userData = { id: user.id, email: user.email, role: user.role };
+			const token = JwtUtility.generateToken(userData);
+			return res.status(200).json({
+				token: token,
+				message: 'Login successful',
+				email: user.email,
+			});
```
Comment (ncutixavier) [inline src/controllers/user.controller.js]: You should print user data after successful login except password

File: src/controllers/user.controller.js [hunk 4]
```diff
@@ -1,22 +1,47 @@
...
+		} catch (err) {
+			return res.status(400).json({ error: err });
+		}
+	}
```
Comment (ncutixavier) [inline src/controllers/user.controller.js]: Follow the same convention implement when registering the user. Check in the controller or ask @Abdden

File: src/middlewares/user.middleware.js [hunk 5]
```diff
@@ -1,12 +1,24 @@
...
+	const { email } = req.body;
+	const user = await User.findOne({ where: { email: email } });
+	if (!user) {
+		return res.status(400).json({ error: 'Wrong credentials' });
+	}
+	if (!user.verified) {
+		return res.status(400).json({ error: 'User email is not verified' });
+	}
+	next();
 };
```
Comment (ncutixavier) [inline src/middlewares/user.middleware.js]: Let's have two separate middleware:
- Check if the user exist in the DB
- Check if the user is verified

File: src/utils/bcrypt.util.js [hunk 6]
```diff
@@ -1,8 +1,13 @@
 import bcrypt from 'bcrypt';
 
+const compare = bcrypt.compare;
```
Comment (ncutixavier) [inline src/utils/bcrypt.util.js]: Since this is not reusable,you declare it where it's being used only

File: tests/user/login.test.js [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+				expect(response.statusCode).toEqual(400);
+			});
+		});
+	});
+	describe('Valid credentials provided', () => {
+		test('token, success message', async () => {
+			const response = await request(app)
+				.post('/api/v1/users/login')
+				.send(verifiedLogin);
+			expect(response.body.message).toEqual('Login successful');
```
Comment (ncutixavier) [inline tests/user/login.test.js]: Can you reduce `describe` in this test? Like putting all tests in a single describe.

File: tests/user/register.test.js [hunk 8]
```diff
@@ -27,12 +27,12 @@ describe('Test User Registration', () => {
     const response = await request(app)
       .post('/api/v1/users/register')
       .send(successRegister);
+      console.log(response.body);
```
Comment (ncutixavier) [inline tests/user/register.test.js]: Remove all logs

**Review by ncutixavier — CHANGES_REQUESTED** [review]

**Review by Abdden — APPROVED** [review]

File: package.json [hunk 9]
```diff
@@ -5,9 +5,9 @@
   "main": "index.js",
   "scripts": {
     "dev": "better-npm-run dev",
-    "migrate": "sequelize db:migrate",
-    "migrate:reset": "sequelize db:migrate:undo:all && npm run migrate",
-    "seed": "sequelize db:seed:all",
+    "migrate": "sequelize db:migrate --name 20230206125837-create-user.js",
+    "migrate:reset": "sequelize db:migrate:undo --name 20230206125837-create-user.js && npm run migrate",
+    "seed": "sequelize db:seed --seed 20230206145743-User.js",
```
Comment (ncutixavier) [inline package.json]: Don't do this. What happens if we have multiple migrations or seedings? Let's use the previous scripts.

File: src/controllers/user.controller.js [hunk 10]
```diff
@@ -2,21 +2,38 @@ import { UserService } from '../services/user.service.js';
...
+	static async loginUser(req, res) {
+		try {
+			const {id, email, firstname, lastname, gender, role, status, avatar, verified} = req.user;
 
+			const userData = { id, email, role, firstname, lastname, gender, status, avatar, verified };
+			const token = JwtUtility.generateToken(userData);
+			return res.status(200).json({
+				token: token,
+				message: 'Login successful',
+				userData,
```
Comment (ncutixavier) [inline src/controllers/user.controller.js]: change the key to be `user`. Hint: `user: userdata`

**Review by ncutixavier — CHANGES_REQUESTED** [review]

File: package.json (hunk 9 above)
Comment (irakozetony) [reply package.json]: For this, circle ci is trying to access other migrations named 'create-jwt-tokens' which are from a different task. Which is causing it to fail!

**Review by ncutixavier — APPROVED** [review]
Good work!

**Pull request merged**